### PR TITLE
C_GetAttributeValue implementation and other improvements

### PR DESCRIPTION
**Pull request by EDVTAZ**
Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

- Implemented C_GetAttributeValue command, but access control needs
to be implemented before this can be used

- Modified code related to C_FindObjectsInit and friends,
so it will match all objects if there are no attributes supplied
in the template, as specified by the standard.

- Added SKS_CKFT_LOGIN_REQUIRED to token flags as default,
because otherwise openssh wouldn't ask for pin number and fail,
since the private key is only available to authenticated users.
I'm not 100% sure that this is the right way to add this flag,
comments are welcome.

- Added automatic CKA_ID generation to the generate key pair
code, because otherwise openssh wouldn't use the generated
key pair. The standard mentions that public and private key
objects may be paired based on their id (although doesn't
require this to be enforced), I assume this is why openssh
only works with key pairs that have CKA_ID.

### CODE REVIEW

**Comment by EDVTAZ** [comment]
Hi @etienne-lms ,
I have been testing the sks with pkcs11-tool and openssh, and made some modifications in it based on that. The details about the changes are in the commit message. Some of the new features I added are not yet complete, but I wanted to show you what I'm working on. Any comments are appreciated. I put a PR on the optee_client repository as well.
One more thing that is not yet in the code because I haven't come up with a solution yet: when openssh starts the authentication with the server it uses the `CKM_RSA_PKCS` mechanism with the `C_Sign` method. Currently this is translated to `TEE_ALG_RSAES_PKCS1_V1_5` on the TEE side, which fails, because this algorithm only supports encryption and decryption. The problem is, however that even though at first this seems easily fixable by translating above mentioned case into `TEE_ALG_RSASSA_PKCS1_V1_5`, it turns out there is no such algorithm in the TEE API, only ones with hash functions integrated, like `TEE_ALG_RSASSA_PKCS1_V1_5_SHA1`. Do you have any ideas on how to solve this problem? Could I maybe use the libtomcrypt library that OPTEE uses in the background?

**Comment by EDVTAZ** [comment]
Right now, I can use ssh with sks backing the client (`ssh -I /usr/lib/libsks.so`). I added the `TEE_ALG_RSASSA_PKCS1_V1_5` to the TEE API, although I don't know if this can be merged. It is not part of the GPD TEE specification, but I don't think it breaks anything either. I also updated the `rsa_sign_hash.c` and `rsa_verify_hash.c` files, because the ones in OP-TEE were outdated, and there was no option to sign without hashing.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c [hunk 1]
```diff
@@ -42,7 +42,6 @@
   @file rsa_sign_hash.c
   RSA PKCS #1 v1.5 and v2 PSS sign hash, Tom St Denis and Andreas Lange
 */
-
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c]: minor: can be discarded.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c [hunk 2]
```diff
@@ -75,15 +74,21 @@ int rsa_sign_hash_ex(const unsigned char *in,       unsigned long  inlen,
...
    /* valid padding? */
-   if ((padding != LTC_PKCS_1_V1_5) && (padding != LTC_PKCS_1_PSS)) {
+   if ((padding != LTC_PKCS_1_V1_5) &&
+       (padding != LTC_PKCS_1_PSS) &&
+       (padding != LTC_PKCS_1_V1_5_NA1)) {
      return CRYPT_PK_INVALID_PADDING;
    }
 
    if (padding == LTC_PKCS_1_PSS) {
-     /* valid prng and hash ? */
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:83]: minor: discard change

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c [hunk 3]
```diff
@@ -75,15 +74,21 @@ int rsa_sign_hash_ex(const unsigned char *in,       unsigned long  inlen,
...
+     /* valid prng ? */
      if ((err = prng_is_valid(prng_idx)) != CRYPT_OK) {
         return err;
      }
+   }
+
+   if (padding != LTC_PKCS_1_V1_5_NA1) {
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:91]: Previous sequence did not check hash for `LTC_PKCS_1_V1_5`. If this is an error, we should fix through a specific commit of propose straight to optee_os.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c [hunk 4]
```diff
@@ -203,6 +218,6 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
 
 #endif /* LTC_MRSA */
 
-/* $Source: /cvs/libtom/libtomcrypt/src/pk/rsa/rsa_verify_hash.c,v $ */
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c]: why this change?

File: lib/libutee/include/tee_api_defines.h [hunk 5]
```diff
@@ -134,6 +134,7 @@
 #define TEE_ALG_DES3_CBC_NOPAD                  0x10000113
 #define TEE_ALG_DES3_CBC_MAC_NOPAD              0x30000113
 #define TEE_ALG_DES3_CBC_MAC_PKCS5              0x30000513
+#define TEE_ALG_RSASSA_PKCS1_V1_5               0x70007830
```
Comment (etienne-lms) [inline lib/libutee/include/tee_api_defines.h]: This would rather go in **tee_api_defines_extensions.h** as [hkdf extensions](https://github.com/OP-TEE/optee_os/blob/master/lib/libutee/include/tee_api_defines_extensions.h#L9).

File: ta_services/secure_key_services/src/processing_asymm.c [hunk 6]
```diff
@@ -138,6 +139,11 @@ static uint32_t sks2tee_algorithm(uint32_t *tee_id,
 		break;
 	}
 
+	if (*tee_id == TEE_ALG_RSAES_PKCS1_V1_5 &&
+		(function == SKS_FUNCTION_SIGN ||
+		 function == SKS_FUNCTION_VERIFY))
+		*tee_id = TEE_ALG_RSASSA_PKCS1_V1_5;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/processing_asymm.c]: I don't think it is nice to switch the algorithm id.
If `TEE_ALG_RSASSA_PKCS1_V1_5` is dedicated to sing/verify a pre-hashed chunk caller should have set `*tee_id` to the right ID.

File: ta_services/secure_key_services/src/pkcs11_attributes.h [hunk 7]
```diff
@@ -11,6 +11,8 @@
 
 #include "serializer.h"
 
+#define CK_UNAVAILABLE_INFORMATION (~0UL)
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/pkcs11_attributes.h]: I found that  `SKS_UNDEFINED_ID` duplicates `CK_UNAVAILABLE_INFORMATION` in the TA API :(

I propose to define  `SKS_UNDEFINED_ID` as an alias to a new `SKS_CK_UNAVAILABLE_INFORMATION` (all `CK_xxx` are prefixed `SKS_` in the TA API, more flexible to prepare integration of the TA in op-tee mainline).

File: ta_services/secure_key_services/src/object.c [hunk 8]
```diff
@@ -731,19 +758,53 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
 			(struct sks_attribute_head *)(void *)cur;
 		len = sizeof(*cli_ref) + cli_ref->size;
 
+		/* check 1. */
+		if (!attribute_is_exportable(cli_ref, obj)) {
+			cli_ref->size = CK_UNAVAILABLE_INFORMATION;
+			attr_sensitive = 1;
+		}
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: I think you should add a `continue;` in that block; don't try to get the data.

File: ta_services/secure_key_services/src/object.c [hunk 9]
```diff
@@ -731,19 +758,53 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
...
 		rv = get_attribute(obj->attributes, cli_ref->id,
 				   cli_ref->size ? cli_ref->data : NULL,
 				   &(cli_ref->size));
+		/* 2. */
+		if (rv == SKS_NOT_FOUND) {
+			cli_ref->size = CK_UNAVAILABLE_INFORMATION;
+			attr_type_invalid = 1;
+		}
+		if (rv == SKS_SHORT_BUFFER)
+			buffer_too_small = 1;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: If `rv` is not `NOT_FOUND`, `SHORT_BUFFER` or `OK`, function should fail straight.

File: ta_services/secure_key_services/src/object.c [hunk 10]
```diff
@@ -725,6 +725,12 @@ uint32_t entry_get_attribute_value(uintptr_t tee_session, TEE_Param *ctrl,
 	if (!obj)
 		return SKS_BAD_PARAM;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: Ouch! Looks like a goto bail is missing here. Can you fix while you're around?

File: ta_services/secure_key_services/src/sks_helpers.c [hunk 11]
```diff
@@ -132,6 +132,24 @@ static const struct processing_id __maybe_unused processing_ids[] = {
 	SKS_PROCESSING_ID(SKS_CKM_SHA384_HMAC),
 	SKS_PROCESSING_ID(SKS_CKM_SHA512_HMAC),
 	SKS_PROCESSING_ID(SKS_CKM_AES_XCBC_MAC),
+	/* Some of these may not be complete, but for testing with
+	 * pkcs11-tool set them to supported TODO G */
+	SKS_PROCESSING_ID(SKS_CKM_RSA_PKCS_KEY_PAIR_GEN),
+	SKS_UNSUPPORTED_PROCESSING_ID(SKS_CKM_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_9796),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_X_509),
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/sks_helpers.c]: this is supported?

File: ta_services/secure_key_services/src/sks_helpers.c [hunk 12]
```diff
@@ -132,6 +132,24 @@ static const struct processing_id __maybe_unused processing_ids[] = {
...
+	SKS_PROCESSING_ID(SKS_CKM_SHA1_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA256_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA384_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA512_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA256_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA384_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA512_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA224_RSA_PKCS),
+	SKS_PROCESSING_ID(SKS_CKM_SHA224_RSA_PKCS_PSS),
+	SKS_PROCESSING_ID(SKS_CKM_RSA_AES_KEY_WRAP),
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/sks_helpers.c]: this is not supported.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c [hunk 13]
```diff
@@ -79,11 +79,12 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
...
 
   if ((padding != LTC_PKCS_1_V1_5) &&
-      (padding != LTC_PKCS_1_PSS)) {
+      (padding != LTC_PKCS_1_PSS) &&
+      (padding != LTC_PKCS_1_V1_5_NA1)) {
     return CRYPT_PK_INVALID_PADDING;
   }
 
-  if (padding == LTC_PKCS_1_PSS) {
+  if (padding != LTC_PKCS_1_V1_5_NA1) {
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:87]: As commented in **rsa_sign_hash.c**: previous sequence did not check hash for `LTC_PKCS_1_V1_5`. If this is an error, we should fix through a specific commit of propose straight to optee_os.

**Review by etienne-lms — COMMENTED** [review]
Thanks for these.
It could be nice to have specific commits for specific changes, but this is not mandatory (time consuming). However, the libtommcrypt change should really be is a dedicated commit and propose straight the optee_os. (does "no ASN.1" really defines a prehashed pkcs1 v1.5  scheme?).

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c (hunk 13 above)
Comment (etienne-lms) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:87]: @EDVTAZ, do you want me to help on this and setup a patch series to propose this to mainline op-tee? This is about signing/verifing a pre-hashed image, right?

Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:87]: I'm sorry for the delay, I was away last week. I can make the pull-request to mainline op-tee, the same way I made the pull requests to your fork, right?

To answer your question, yes it is about signing/verifying a pre-hashed image. The only problem was that there was no such operation in the GPD TEE API, that is why I added TEE_ALG_RSASSA_PKCS1_V1_5.

I will go through your other comments too now.

File: ta_services/secure_key_services/src/sks_helpers.c (hunk 11 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/sks_helpers.c]: I just added these so I can find what mechanisms client applications needed. I will look through them more carefully and remove the ones that are incorrect.

File: ta_services/secure_key_services/src/sks_helpers.c (hunk 12 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/sks_helpers.c]: I will remove it, thanks for pointing it out.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 1 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c]: I replaced this file with a newer version, that's the reason these comments are a bit different. I will go through it again and discard these changes.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 2 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:83]: Same as the other one, I will discard.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c (hunk 4 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c]: I will discard.

File: lib/libutee/include/tee_api_defines.h (hunk 5 above)
Comment (EDVTAZ) [reply lib/libutee/include/tee_api_defines.h]: I wasn't aware of that, I will move it.

File: ta_services/secure_key_services/src/processing_asymm.c (hunk 6 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/processing_asymm.c]: This is part of the function that translates the SKS_CKM_* to tee_id, and SKS_CKM_RSA_PKCS can be either TEE_ALG_RSAES_PKCS1_V1_5 or TEE_ALG_RSASSA_PKCS1_V1_5 based on whether it is called with C_Encrypt/C_Decrypt or C_Sign/C_Verify. I will place this chunk closer to the code that does the initial translation, that way it will be more intuitive.

File: ta_services/secure_key_services/src/object.c (hunk 8 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/object.c]: Yes, I will do that.

File: ta_services/secure_key_services/src/object.c (hunk 9 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/object.c]: The pkcs11 standard asks that every attribute should be processed:

> Note that the error codes CKR_ATTRIBUTE_SENSITIVE, CKR_ATTRIBUTE_TYPE_INVALID, and CKR_BUFFER_TOO_SMALL do not denote true errors for C_GetAttributeValue. If a call to C_GetAttributeValue returns any of these three values, then the call MUST nonetheless have processed every attribute in the template supplied to C_GetAttributeValue. Each attribute in the template whose value can be returned by the call to C_GetAttributeValue will be returned by the call to C_GetAttributeValue.

File: ta_services/secure_key_services/src/pkcs11_attributes.h (hunk 7 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/pkcs11_attributes.h]: Ok, I think I can do that.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 2 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:83]: Actually I'm not sure about this one. If I understand this correctly, this block only checks the prng, the next one checks hash for every padding (except for LTC_PKCS_1_V1_5_NA1, where there is no need for a hash). It seems to me that the older version of the comment is incorrect.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 3 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:91]: I think this is correct the way it is, the hash for `LTC_PKCS_1_V1_5` is being checked in this block. Also, see my comment above.

**Comment by EDVTAZ** [comment]
Hi @etienne-lms ,
Thank you for your comments. If you agree with my comments about libtomcrypt, I can make a PR to mainline OP-TEE.

Regarding `LTC_PKCS1_V1_5_NA1`, I looked into it again, and it seems there was a bit of a misunderstanding on my part. None of the `TEE_AsymmetricSignDigest` operations actually do the hashing operation, they expect the data to be pre-hashed. However they need to know the hash function used because they include that information in the padding. However with `CKM_RSA_PKCS` we need `EMSA` padding (https://lists.randombit.net/pipermail/botan-devel/2008-November/000697.html), and `LTC_PKCS1_V1_5_NA1` does that.  
This means that currently all `CKM_*_RSA_PKCS` mechanisms in sks are missing a digest operation, because those mechanisms do hashing and signing too. I will be looking into that next.

And one last thing, I also added a PR for client side: https://github.com/etienne-lms/optee_client/pull/10. If you have time you could take a look at that as well. Thanks!

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 3 above)
Comment (etienne-lms) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:91]: my fault. you're right.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c (hunk 2 above)
Comment (etienne-lms) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c:83]: sorry. ok with your change.

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c (hunk 13 above)
Comment (etienne-lms) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:87]: i think it would be nice pushing a P-R for this libutee and libtomcrypt change in the mainline OP-TEE.
Maintainer and I will help finding the best way to get this pre-hashed scheme supported.

File: ta_services/secure_key_services/include/sks_ta.h [hunk 14]
```diff
@@ -709,13 +709,15 @@ struct sks_attribute_head {
...
+#define SKS_CKR_ATTRIBUTE_SENSITIVE		0x0000002d
 
 /* Status without strict equivalence in Cryptoki API */
 #define SKS_NOT_FOUND				0x00001000
 #define SKS_NOT_IMPLEMENTED			0x00001001
 
 /* Attribute specific values */
-#define SKS_UNDEFINED_ID			((uint32_t)0xFFFFFFFF)
+#define SKS_CK_UNAVAILABLE_INFORMATION		((uint32_t)0xFFFFFFFF)
+#define SKS_UNDEFINED_ID			SKS_CK_UNAVAILABLE_INFORMATION
```
Comment (etienne-lms) [inline ta_services/secure_key_services/include/sks_ta.h:720]: thanks

File: ta_services/secure_key_services/src/object.c [hunk 15]
```diff
@@ -521,11 +521,15 @@ uint32_t entry_find_objects_init(uintptr_t tee_session, TEE_Param *ctrl,
 		uint32_t obj_handle;
 		uint32_t *handles;
 
-		rv = token_obj_matches_ref(req_attrs, obj);
-		if (rv == SKS_NOT_FOUND)
-			continue;
-		if (rv != SKS_OK)
-			goto bail;
+		/* If there are no attributes specified, we return
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: minor: let's use the op-tee coding style. multi-line comment go
```c
/*
 * <start comment from here>
 */
```

File: ta_services/secure_key_services/src/object.c [hunk 16]
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
...
+
+	session = sks_handle2session(session_handle, tee_session);
+	if (!session) {
+		rv = SKS_CKR_SESSION_HANDLE_INVALID;
+		goto bail;
+	}
+
+	obj = sks_handle2object(object_handle, session);
+	if (!obj)
+		return SKS_BAD_PARAM;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: s/`return SKS_...`/`rv = SKS_...; goto bail;`/

File: ta_services/secure_key_services/src/object.c [hunk 17]
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
...
+	if (buffer_too_small)
+		rv = SKS_CKR_BUFFER_TOO_SMALL;
+	
+	/* Move updated template to out buffer */
+	TEE_MemMove(out->memref.buffer, template, out->memref.size);
+
+	TEE_Free(template);
+	template = NULL;
+
+bail:
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: move `bail:` before `TEE_Free(template);`.

File: ta_services/secure_key_services/src/object.c [hunk 18]
```diff
@@ -673,3 +677,141 @@ uint32_t entry_find_objects_final(uintptr_t tee_session, TEE_Param *ctrl,
...
+		rv = get_attribute(obj->attributes, cli_ref->id,
+				   cli_ref->size ? cli_ref->data : NULL,
+				   &(cli_ref->size));
+		/* 2. */
+		if (rv == SKS_NOT_FOUND) {
+			cli_ref->size = SKS_CK_UNAVAILABLE_INFORMATION;
+			attr_type_invalid = 1;
+		}
+		if (rv == SKS_SHORT_BUFFER)
+			buffer_too_small = 1;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/object.c]: I think sequence should be halted if `rc` is non of the above neither `SKS_OK`.

File: ta_services/secure_key_services/src/pkcs11_attributes.c [hunk 19]
```diff
@@ -1485,3 +1485,50 @@ bool object_is_private(struct sks_attrs_head *head)
...
+
+	/* Check if SKS_CKA_ID is already present */
+	rv = get_attribute(*attrs, SKS_CKA_ID, NULL, NULL);
+	if (rv != SKS_NOT_FOUND)
+		return SKS_OK;
+	
+	/* Generate 32 bit ID */
+	if (!set_id)
+		TEE_GenerateRandom(&id, sizeof(id));
+	else id = *set_id;
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/pkcs11_attributes.c]: minor: split in 2 lines:
```c
else
	id = *set_id;
```

File: ta_services/secure_key_services/src/pkcs11_attributes.c [hunk 20]
```diff
@@ -1485,3 +1485,50 @@ bool object_is_private(struct sks_attrs_head *head)
...
+	uint32_t id;
+
+	/* Check if SKS_CKA_ID is already present */
+	rv = get_attribute(*attrs, SKS_CKA_ID, NULL, NULL);
+	if (rv != SKS_NOT_FOUND)
+		return SKS_OK;
+	
+	/* Generate 32 bit ID */
+	if (!set_id)
+		TEE_GenerateRandom(&id, sizeof(id));
```
Comment (etienne-lms) [inline ta_services/secure_key_services/src/pkcs11_attributes.c]: no issue with collision? should it insure ID unique?

File: ta_services/secure_key_services/src/processing_asymm.c (hunk 6 above)
Comment (etienne-lms) [reply ta_services/secure_key_services/src/processing_asymm.c]: ok, fine. I did not understand the sequence at the time.

**Comment by etienne-lms** [comment]
Nice changes. sorry for the late feedback.
Do you think you can rebuild this patch serie with specific topic changes?

File: core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c (hunk 13 above)
Comment (EDVTAZ) [reply core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:87]: Ok, I will do that.

File: ta_services/secure_key_services/src/object.c (hunk 18 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/object.c]: Ok, I will correct this.

File: ta_services/secure_key_services/src/pkcs11_attributes.c (hunk 20 above)
Comment (EDVTAZ) [reply ta_services/secure_key_services/src/pkcs11_attributes.c]: I added this, so OpenSSH would operate on the keys I generated (since it would refuse if there was no CKA_ID). It would be nice to check for collisions, but at the time I was writing this I couldn't come up with any easy way to do so and later forgot about it. The standard also doesn't seem to require us to ensure this:
> The CKA_ID attribute is intended as a means of distinguishing multiple public-key/private-key pairs held by the same subject (whether stored in the same token or not). (Since the keys are distinguished by subject name as well as identifier, it is possible that keys for different subjects may have the same CKA_ID value without introducing any ambiguity.)
> It is intended in the interests of interoperability that the subject name and key identifier for a certificate will be the same as those for the corresponding public and private keys (though it is not required that all be stored in the same token). However, Cryptoki does not enforce this association, or even the uniqueness of the key identifier for a given subject; in particular, an application may leave the key identifier empty.

In any case, I will look into how I can iterate through all objects present in the SKS and implement this check.

**Comment by EDVTAZ** [comment]
Thanks, I will rebuild it once I'm done with the cka_id collision detection, if that's all right with you. I will also make the P-R to mainline OPTEE. University started for me, so things may go a bit slower, but I'm working on it.

**Comment by etienne-lms** [comment]
Thanks for your work. I think we can drop the CKA_ID collision issue. As your pointed at, collision are not forbidden and we cannot check against outer tokens :)
So a 32bit random is fine. To avoid collision, one may want 128bit UUID.

**Comment by etienne-lms** [comment]
merged through https://github.com/etienne-lms/optee_os/pull/12.
